crates/uv/tests/lock_scenarios.rs

```diff
@@ -0,0 +1,803 @@
+          And because only project==0.1.0 is available and project depends on project, we can conclude that the requirements are unsatisfiable.
+    "###
+    );
+
+    Ok(())
+}
+
+/// This setup introduces dependencies on two distinct versions of `c`, where each
+/// such dependency has a marker expression attached that would normally make them
+/// disjoint. In a non-universal resolver, this is no problem. But in a forking
+/// resolver that tries to create one universal resolution, this can lead to two
+/// distinct versions of `c` in the resolution. This is in and of itself not a
+/// problem, since that is an expected scenario for universal resolution. The
+/// problem in this case is that because the dependency specifications for `c` occur
+/// in two different points (i.e., they are not sibling dependency specifications)
+/// in the dependency graph, the forking resolver does not "detect" it, and thus
+/// never forks and thus this results in "no resolution."
+///
+/// ```text
+/// fork-non-local-fork-marker-transitive
+/// ├── environment
+/// │   └── python3.8
+/// ├── root
+/// │   ├── requires a==1.0.0
+/// │   │   └── satisfied by a-1.0.0
+/// │   └── requires b==1.0.0
+/// │       └── satisfied by b-1.0.0
+/// ├── a
+/// │   └── a-1.0.0
+/// │       └── requires c<2.0.0; sys_platform == "linux"
+/// │           └── satisfied by c-1.0.0
+/// ├── b
+/// │   └── b-1.0.0
+/// │       └── requires c>=2.0.0; sys_platform == "darwin"
+/// │           └── satisfied by c-2.0.0
+/// └── c
+///     ├── c-1.0.0
+///     └── c-2.0.0
+/// ```
+#[test]
+fn fork_non_local_fork_marker_transitive() -> Result<()> {
+    let context = TestContext::new("3.8");
+
+    // In addition to the standard filters, swap out package names for shorter messages
+    let mut filters = context.filters();
+    filters.push((r"fork-non-local-fork-marker-transitive-", "package-"));
+
+    let pyproject_toml = context.temp_dir.child("pyproject.toml");
+    pyproject_toml.write_str(
+        r###"
+        [project]
+        name = "project"
+        version = "0.1.0"
+        dependencies = [
+          '''fork-non-local-fork-marker-transitive-a==1.0.0''',
+          '''fork-non-local-fork-marker-transitive-b==1.0.0''',
+        ]
+        "###,
+    )?;
+
+    let mut cmd = context.lock_without_exclude_newer();
+    cmd.arg("--index-url")
+        .arg("https://astral-sh.github.io/packse/0.3.17/simple-html/");
+    uv_snapshot!(filters, cmd, @r###"
+    success: false
+    exit_code: 1
+    ----- stdout -----
+
+    ----- stderr -----
+    warning: `uv lock` is experimental and may change without warning.
+      × No solution found when resolving dependencies:
+      ╰─▶ Because package-b==1.0.0 depends on package-c{sys_platform == 'darwin'}>=2.0.0 and only package-c{sys_platform == 'darwin'}<=2.0.0 is available, we can conclude that package-b==1.0.0 depends on package-c{sys_platform == 'darwin'}==2.0.0.
```

Comment:
Oof these markers are tough on error messages

Reply:
Yeah indeed. We could have an alternative display format that excludes them in error messages, but they also simultaneously seem useful... Not quite sure what to do about that yet.